### PR TITLE
feat: created a component for apdex

### PR DESCRIPTION
**Pull request by Rajat-Dabade**
This PR adds an ApDex graph in the Application Layer.

Close https://github.com/SigNoz/signoz/issues/1477

Loom video:  https://www.loom.com/share/99e60110575342d3ac391252f38b2c17?sid=679c76d4-13c7-41da-9848-08d04926b21e

### CODE REVIEW

File: frontend/src/constants/theme.ts [hunk 1]
```diff
@@ -52,6 +52,7 @@ const themeColors = {
 	gamboge: '#D89614',
 	bckgGrey: '#1d1d1d',
 	lightBlue: '#177ddc',
+	lightGrey: '#acacac',
```
Comment (palashgdev) [inline frontend/src/constants/theme.ts:55]: grey is already present in color from antd/colors

File: frontend/src/container/GridGraphLayout/WidgetHeader/DisplayThreshold.tsx [hunk 2]
```diff
@@ -0,0 +1,13 @@
+import { DisplayThresholdContainer, Typography } from './styles';
+import { DisplayThresholdProps } from './types';
+
+function DisplayThreshold({ threshold }: DisplayThresholdProps): JSX.Element {
+	return (
+		<DisplayThresholdContainer>
+			<div>Threshold </div>
```
Comment (palashgdev) [inline frontend/src/container/GridGraphLayout/WidgetHeader/DisplayThreshold.tsx:7]: Typography

File: frontend/src/container/MetricsApplication/MetricsPageQueries/OverviewQueries.ts [hunk 3]
```diff
@@ -85,6 +91,365 @@ export const latency = ({
...
+		{
+			id: '',
+			key: {
+				key: WidgetKeys.DurationNano,
+				dataType: DataType.FLOAT64,
+				isColumn: true,
+				type: MetricsType.Tag,
+			},
+			op: OPERATORS['<='],
+			value: threashold * 1000000000,
```
Comment (palashgdev) [inline frontend/src/container/MetricsApplication/MetricsPageQueries/OverviewQueries.ts:168]: util for this one

File: frontend/src/container/MetricsApplication/Tabs/Overview.tsx [hunk 4]
```diff
@@ -225,25 +227,45 @@ function Application(): JSX.Element {
...
-						opName="Error"
+						topLevelOperationsRoute={topLevelOperationsRoute}
+						tagFilterItems={tagFilterItems}
 					/>
+					<ColContainer>
+						<Button
+							type="default"
+							size="small"
+							id="Error_button"
+							onClick={(): void => {
```
Comment (palashgdev) [inline frontend/src/container/MetricsApplication/Tabs/Overview.tsx:250]: closure

File: frontend/src/container/MetricsApplication/Tabs/Overview/ApDex/ApDexMetrics.tsx [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+		query: {
+			queryType: EQueryType.QUERY_BUILDER,
+			promql: [],
+			builder: apDexMetricsQueryBuilderQueries({
+				servicename,
+				tagFilterItems,
+				topLevelOperationsRoute,
+				threashold: thresholdValue || 0,
+				delta: delta || false,
+				le: le || [],
```
Comment (palashgdev) [inline frontend/src/container/MetricsApplication/Tabs/Overview/ApDex/ApDexMetrics.tsx:34]: can we better naming here

File: frontend/src/container/MetricsApplication/Tabs/Overview/ApDex/ApDexMetricsApplication.tsx [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+	thresholdValue,
+}: ApDexDataSwitcherProps): JSX.Element {
+	const { data, isLoading, error } = useGetMetricMeta(metricMeta);
+	const { notifications } = useNotifications();
+
+	if (isLoading) {
+		return <Spinner height="40vh" tip="Loading..." />;
+	}
+
+	if (error && axios.isAxiosError(error)) {
```
Comment (palashgdev) [inline frontend/src/container/MetricsApplication/Tabs/Overview/ApDex/ApDexMetricsApplication.tsx:24]: we already have hook

File: frontend/src/container/MetricsApplication/Tabs/Overview/ApDex/ApDexTraces.tsx [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+import { ApDexDataSwitcherProps } from './types';
+
+function ApDexTraces({
+	handleGraphClick,
+	onDragSelect,
+	topLevelOperationsRoute,
+	tagFilterItems,
+	thresholdValue,
+}: ApDexDataSwitcherProps): JSX.Element {
+	console.log('thresholdValue', thresholdValue);
```
Comment (palashgdev) [inline frontend/src/container/MetricsApplication/Tabs/Overview/ApDex/ApDexTraces.tsx:20]: log

File: frontend/src/container/MetricsApplication/Tabs/Overview/ApDex/index.tsx [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+	topLevelOperationsRoute,
+	tagFilterItems,
+}: ApDexApplicationProps): JSX.Element {
+	const { servicename } = useParams<IServiceName>();
+	const { data, isLoading, error, isRefetching } = useGetApDexSettings(
+		servicename,
+	);
+	const { notifications } = useNotifications();
+
+	if (error && axios.isAxiosError(error)) {
```
Comment (palashgdev) [inline frontend/src/container/MetricsApplication/Tabs/Overview/ApDex/index.tsx:25]: same here

File: frontend/src/container/MetricsApplication/Tabs/Overview/ApDex/index.tsx [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+		notifications.error({
+			message: error.message,
+		});
+	}
+
+	return (
+		<Card>
+			<GraphContainer>
+				{isLoading || isRefetching ? (
```
Comment (palashgdev) [inline frontend/src/container/MetricsApplication/Tabs/Overview/ApDex/index.tsx:34]: we can have condition to show the spinner above

File: frontend/src/container/MetricsApplication/Tabs/Overview/ApDex/types.ts [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+
+export interface ApDexApplicationProps {
+	handleGraphClick: (type: string) => ClickHandlerType;
+	onDragSelect: (start: number, end: number) => void;
+	topLevelOperationsRoute: string[];
+	tagFilterItems: TagFilterItem[];
+}
+
+export interface ApDexDataSwitcherProps extends ApDexApplicationProps {
+	thresholdValue: number | undefined;
```
Comment (palashgdev) [inline frontend/src/container/MetricsApplication/Tabs/Overview/ApDex/types.ts:13]:
```suggestion
	thresholdValue?: number;
```

File: frontend/src/container/MetricsApplication/Tabs/Overview/ApDex/types.ts [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+}
+
+export interface ApDexMetricsProps extends ApDexDataSwitcherProps {
+	delta?: boolean;
+	le?: number[];
```
Comment (palashgdev) [inline frontend/src/container/MetricsApplication/Tabs/Overview/ApDex/types.ts:18]: same here

File: frontend/src/hooks/apDex/useGetMetricMeta.ts [hunk 12]
```diff
@@ -0,0 +1,14 @@
+import { getMetricMeta } from 'api/metrics/ApDex/getMetricMeta';
+import { AxiosResponse } from 'axios';
+import { useQuery, UseQueryResult } from 'react-query';
+import { MetricMetaProps } from 'types/api/metrics/getApDex';
+
+export const useGetMetricMeta = (
+	metricName: string,
+): UseQueryResult<AxiosResponse<MetricMetaProps>> => {
+	const queryKey = [{ metricName }];
```
Comment (palashgdev) [inline frontend/src/hooks/apDex/useGetMetricMeta.ts:9]: we can directly use the queryKey

File: frontend/src/hooks/apDex/useGetApDexSettings.ts [hunk 13]
```diff
@@ -0,0 +1,14 @@
+import { getApDexSettings } from 'api/metrics/ApDex/getApDexSettings';
+import { AxiosResponse } from 'axios';
+import { useQuery, UseQueryResult } from 'react-query';
+import { ApDexPayloadProps } from 'types/api/metrics/getApDex';
+
+export const useGetApDexSettings = (
+	servicename: string,
+): UseQueryResult<AxiosResponse<ApDexPayloadProps[]>> => {
+	const queryKey = [{ servicename }];
```
Comment (palashgdev) [inline frontend/src/hooks/apDex/useGetApDexSettings.ts:9]: same here

File: frontend/src/hooks/apDex/useSetApDexSettings.ts [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+export const useSetApDexSettings = ({
+	servicename,
+	threshold,
+	excludeStatusCode,
+}: SetApDexSettingsProps): UseMutationResult<
+	SetApDexPayloadProps,
+	Error,
+	SetApDexSettingsProps
+> => {
+	const queryKey = [servicename, threshold.toString(), excludeStatusCode];
```
Comment (palashgdev) [inline frontend/src/hooks/apDex/useSetApDexSettings.ts:17]: here also

File: frontend/src/pages/MetricsApplication/ApDex/ApDexApplication.tsx [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+
+	const handlePopOverClose = (): void => {
+		setIsOpen(false);
+	};
+
+	const handleOpenChange = (newOpen: boolean): void => {
+		setIsOpen(newOpen);
+	};
+
+	if (error && axios.isAxiosError(error)) {
```
Comment (palashgdev) [inline frontend/src/pages/MetricsApplication/ApDex/ApDexApplication.tsx:26]: use Hook for this one

File: frontend/src/pages/MetricsApplication/ApDex/ApDexSettings.tsx [hunk 16]
```diff
@@ -0,0 +1,124 @@
...
+	const [thresholdValue, setThresholdValue] = useState(0);
+	const { notifications } = useNotifications();
+
+	const { isLoading: setApDexIsLoading, mutateAsync } = useSetApDexSettings({
+		servicename,
+		threshold: thresholdValue,
+		excludeStatusCode: '',
+	});
+
+	useEffect(() => {
```
Comment (palashgdev) [inline frontend/src/pages/MetricsApplication/ApDex/ApDexSettings.tsx:37]: why useEffect here can use the initial callback of useState ?

File: frontend/src/constants/theme.ts (hunk 1 above)
Comment (Rajat-Dabade) [reply frontend/src/constants/theme.ts:55]: Okay

File: frontend/src/container/GridGraphLayout/WidgetHeader/DisplayThreshold.tsx (hunk 2 above)
Comment (Rajat-Dabade) [reply frontend/src/container/GridGraphLayout/WidgetHeader/DisplayThreshold.tsx:7]: Okay

File: frontend/src/container/MetricsApplication/MetricsPageQueries/OverviewQueries.ts (hunk 3 above)
Comment (Rajat-Dabade) [reply frontend/src/container/MetricsApplication/MetricsPageQueries/OverviewQueries.ts:168]: Okay

File: frontend/src/container/MetricsApplication/Tabs/Overview.tsx (hunk 4 above)
Comment (Rajat-Dabade) [reply frontend/src/container/MetricsApplication/Tabs/Overview.tsx:250]: Okay

File: frontend/src/container/MetricsApplication/Tabs/Overview/ApDex/ApDexMetrics.tsx (hunk 5 above)
Comment (Rajat-Dabade) [reply frontend/src/container/MetricsApplication/Tabs/Overview/ApDex/ApDexMetrics.tsx:34]: Okay

File: frontend/src/container/MetricsApplication/Tabs/Overview/ApDex/ApDexMetricsApplication.tsx (hunk 6 above)
Comment (Rajat-Dabade) [reply frontend/src/container/MetricsApplication/Tabs/Overview/ApDex/ApDexMetricsApplication.tsx:24]: Okay

File: frontend/src/container/MetricsApplication/Tabs/Overview/ApDex/ApDexTraces.tsx (hunk 7 above)
Comment (Rajat-Dabade) [reply frontend/src/container/MetricsApplication/Tabs/Overview/ApDex/ApDexTraces.tsx:20]: Removed it.

File: frontend/src/container/MetricsApplication/Tabs/Overview/ApDex/index.tsx (hunk 9 above)
Comment (Rajat-Dabade) [reply frontend/src/container/MetricsApplication/Tabs/Overview/ApDex/index.tsx:34]: Okay

File: frontend/src/container/MetricsApplication/Tabs/Overview/ApDex/index.tsx (hunk 8 above)
Comment (Rajat-Dabade) [reply frontend/src/container/MetricsApplication/Tabs/Overview/ApDex/index.tsx:25]: Okay

File: frontend/src/container/MetricsApplication/Tabs/Overview/ApDex/types.ts (hunk 11 above)
Comment (Rajat-Dabade) [reply frontend/src/container/MetricsApplication/Tabs/Overview/ApDex/types.ts:18]: Changed it.

File: frontend/src/hooks/apDex/useGetApDexSettings.ts (hunk 13 above)
Comment (Rajat-Dabade) [reply frontend/src/hooks/apDex/useGetApDexSettings.ts:9]: Sure.

File: frontend/src/hooks/apDex/useSetApDexSettings.ts (hunk 14 above)
Comment (Rajat-Dabade) [reply frontend/src/hooks/apDex/useSetApDexSettings.ts:17]: okay

File: frontend/src/pages/MetricsApplication/ApDex/ApDexApplication.tsx (hunk 15 above)
Comment (Rajat-Dabade) [reply frontend/src/pages/MetricsApplication/ApDex/ApDexApplication.tsx:26]: Okay

File: frontend/src/pages/MetricsApplication/ApDex/ApDexSettings.tsx (hunk 16 above)
Comment (Rajat-Dabade) [reply frontend/src/pages/MetricsApplication/ApDex/ApDexSettings.tsx:37]: Okay.

File: frontend/src/hooks/apDex/useGetMetricMeta.ts (hunk 12 above)
Comment (Rajat-Dabade) [reply frontend/src/hooks/apDex/useGetMetricMeta.ts:9]: Done.

**Comment by srikanthccv** [comment]
Screenshot 2023-08-09 at 3 47 28 PM

- threshold can be greater than 1 because some applications can have high latency
Screenshot 2023-08-09 at 3 48 35 PM

- AppDex Threshold -> Apdex threshold (in seconds) and info icon is not working.


- I noticed there is an issue in the backend when we use traces payload. The issue is when there are no spans < threshold but there are spans for 4 * threshold, the charts show no data (For example, check frontend service). I am thinking we should always rely on metrics whether the `USE_SPAN_METRICS` flag is set or not (we will enable it again later for traces when things are sorted on the backend). Is this a small change? I will confirm with Pranay and share if we want to go this route.

**Comment by Rajat-Dabade** [comment]
@srikanthccv, it's a small change actually, we can use metrics data if we want to.

File: frontend/src/container/MetricsApplication/Tabs/Overview/ApDex/ApDexMetrics.tsx [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+	metricsBuckets,
+	thresholdValue,
+	onDragSelect,
+	tagFilterItems,
+	topLevelOperationsRoute,
+	handleGraphClick,
+}: ApDexMetricsProps): JSX.Element {
+	const { servicename } = useParams<IServiceName>();
+
+	const apDexMetricsWidget = getWidgetQueryBuilder({
```
Comment (palashgdev) [inline frontend/src/container/MetricsApplication/Tabs/Overview/ApDex/ApDexMetrics.tsx:24]: memo

File: frontend/src/container/MetricsApplication/Tabs/Overview/ApDex/ApDexTraces.tsx [hunk 18]
```diff
@@ -0,0 +1,59 @@
...
+function ApDexTraces({
+	handleGraphClick,
+	onDragSelect,
+	topLevelOperationsRoute,
+	tagFilterItems,
+	thresholdValue,
+}: ApDexDataSwitcherProps): JSX.Element {
+	const { servicename } = useParams<IServiceName>();
+
+	const apDexTracesWidget = getWidgetQueryBuilder({
```
Comment (palashgdev) [inline frontend/src/container/MetricsApplication/Tabs/Overview/ApDex/ApDexTraces.tsx:22]: same here

File: frontend/src/container/MetricsApplication/Tabs/Overview/ApDex/index.tsx [hunk 19]
```diff
@@ -0,0 +1,43 @@
...
+}: ApDexApplicationProps): JSX.Element {
+	const { servicename } = useParams<IServiceName>();
+	const { data, isLoading, error, isRefetching } = useGetApDexSettings(
+		servicename,
+	);
+	useErrorNotification(error);
+
+	return (
+		<Card>
+			{isLoading || isRefetching ? (
```
Comment (palashgdev) [inline frontend/src/container/MetricsApplication/Tabs/Overview/ApDex/index.tsx:26]: loading condition should be handle above

File: frontend/src/pages/MetricsApplication/utils.ts [hunk 20]
```diff
@@ -15,3 +18,41 @@ export const getMetricsApplicationKey = (
...
+	servicename,
+}: OnSaveApDexSettingsProps) => (): void => {
+	if (thresholdValue > 0 && thresholdValue < 1 && refetch) {
+		mutateAsync({
+			servicename,
+			threshold: thresholdValue,
+			excludeStatusCode: '',
+		})
+			.then(() => {
+				refetch();
```
Comment (palashgdev) [inline frontend/src/pages/MetricsApplication/utils.ts:37]: what we are referencing here can we have a better name here

File: frontend/src/types/api/queryBuilder/queryBuilderData.ts [hunk 21]
```diff
@@ -20,7 +20,7 @@ export interface TagFilterItem {
 	id: string;
 	key?: BaseAutocompleteData;
 	op: string;
-	value: string[] | string;
+	value: string[] | string | number[] | number | boolean[] | boolean;
```
Comment (palashgdev) [inline frontend/src/types/api/queryBuilder/queryBuilderData.ts:23]: do we have such inconsistency ?

File: frontend/src/container/MetricsApplication/Tabs/Overview/ApDex/ApDexMetrics.tsx (hunk 17 above)
Comment (Rajat-Dabade) [reply frontend/src/container/MetricsApplication/Tabs/Overview/ApDex/ApDexMetrics.tsx:24]: Okay

File: frontend/src/container/MetricsApplication/Tabs/Overview/ApDex/ApDexTraces.tsx (hunk 18 above)
Comment (Rajat-Dabade) [reply frontend/src/container/MetricsApplication/Tabs/Overview/ApDex/ApDexTraces.tsx:22]: Okay

File: frontend/src/container/MetricsApplication/Tabs/Overview/ApDex/index.tsx (hunk 19 above)
Comment (Rajat-Dabade) [reply frontend/src/container/MetricsApplication/Tabs/Overview/ApDex/index.tsx:26]: Okay.

File: frontend/src/pages/MetricsApplication/utils.ts (hunk 20 above)
Comment (Rajat-Dabade) [reply frontend/src/pages/MetricsApplication/utils.ts:37]: Okay.

File: frontend/src/types/api/queryBuilder/queryBuilderData.ts (hunk 21 above)
Comment (Rajat-Dabade) [reply frontend/src/types/api/queryBuilder/queryBuilderData.ts:23]: As of now we have `string[] | string | number | boolean`, `number[]` and `boolean[]` was added by mistake,  I will remove it.

**Comment by srikanthccv** [comment]
Please resolve the conflicts and use the metrics instead of relying on `USE_SPAN_METRICS` flag.

**Comment by srikanthccv** [comment]
https://www.loom.com/share/c838ee58aea54f83addc9d33e0dfb119

File: frontend/src/api/metrics/ApDex/getApDexSettings.ts [hunk 22]
```diff
@@ -0,0 +1,8 @@
+import axios from 'api';
+import { AxiosResponse } from 'axios';
+import { ApDexPayloadProps } from 'types/api/metrics/getApDex';
+
+export const getApDexSettings = (
+	servicename: string,
+): Promise<AxiosResponse<ApDexPayloadProps[]>> =>
+	axios.get(`settings/apdex?services=${servicename}`);
```
Comment (palashgdev) [inline frontend/src/api/metrics/ApDex/getApDexSettings.ts:8]:
```suggestion
	axios.get(`/settings/apdex?services=${servicename}`);
```

File: frontend/src/container/GridGraphLayout/Graph/WidgetGraphComponent.tsx [hunk 23]
```diff
@@ -53,6 +53,8 @@ function WidgetGraphComponent({
 	setLayout,
 	onDragSelect,
 	onClickHandler,
+	allowThreshold = false,
```
Comment (palashgdev) [inline frontend/src/container/GridGraphLayout/Graph/WidgetGraphComponent.tsx:56]: right content should be node

File: frontend/src/container/MetricsApplication/MetricsPageQueries/OverviewQueries.ts [hunk 24]
```diff
@@ -85,6 +91,365 @@ export const latency = ({
...
+		},
+		{
+			id: '',
+			key: {
+				key: WidgetKeys.Name,
+				dataType: DataType.STRING,
+				isColumn: true,
+				type: MetricsType.Tag,
+			},
+			op: OPERATORS.IN.toLowerCase(), // TODO: need to remove toLowerCase() this once backend is changed
```
Comment (palashgdev) [inline frontend/src/container/MetricsApplication/MetricsPageQueries/OverviewQueries.ts:189]: now this `toLowerCase` is not needed

File: frontend/src/container/MetricsApplication/MetricsPageQueries/OverviewQueries.ts [hunk 25]
```diff
@@ -85,6 +91,365 @@ export const latency = ({
...
```
Comment (palashgdev) [inline frontend/src/container/MetricsApplication/MetricsPageQueries/OverviewQueries.ts:236]: same here

File: frontend/src/container/MetricsApplication/Tabs/Overview/ApDex/ApDexMetrics.tsx [hunk 26]
```diff
@@ -0,0 +1,94 @@
...
+					}),
+					clickhouse_sql: [],
+					id: uuid(),
+				},
+				title: (
+					<Typography>
+						{GraphTitle.APDEX}
+						<Typography.Link>
+							<TextToolTip
+								text="Apdex is a way to measure your users' satisfaction with the response time of your web service. It's represented as a score from 0-1."
```
Comment (palashgdev) [inline frontend/src/container/MetricsApplication/Tabs/Overview/ApDex/ApDexMetrics.tsx:49]: constant

File: frontend/src/pages/MetricsApplication/ApDex/ApDexApplication.tsx [hunk 27]
```diff
@@ -0,0 +1,55 @@
...
+import { useGetApDexSettings } from 'hooks/apDex/useGetApDexSettings';
+import useErrorNotification from 'hooks/useErrorNotification';
+import { useState } from 'react';
+import { useParams } from 'react-router-dom';
+
+import { Button } from '../styles';
+import ApDexSettings from './ApDexSettings';
+
+function ApDexApplication(): JSX.Element {
+	const { servicename } = useParams<{ servicename: string }>();
```
Comment (palashgdev) [inline frontend/src/pages/MetricsApplication/ApDex/ApDexApplication.tsx:12]:
```suggestion
	const { servicename } = useParams<IServiceName>();
```

File: frontend/src/pages/MetricsApplication/ApDex/ApDexSettings.tsx [hunk 28]
```diff
@@ -0,0 +1,115 @@
+/* eslint-disable react/jsx-no-comment-textnodes */
```
Comment (palashgdev) [inline frontend/src/pages/MetricsApplication/ApDex/ApDexSettings.tsx:1]: please remove this eslint disabling

File: frontend/src/pages/MetricsApplication/ApDex/ApDexSettings.tsx [hunk 29]
```diff
@@ -0,0 +1,115 @@
...
+						Save
+					</SaveButton>
+				</SaveAndCancelContainer>,
+			]}
+		>
+			<AppDexThresholdContainer>
+				<Typography>
+					Apdex threshold (in seconds){' '}
+					<TextToolTip
+						text="Apdex is a way to measure your users' satisfaction with the response time of your web service. It's represented as a score from 0-1."
```
Comment (palashgdev) [inline frontend/src/pages/MetricsApplication/ApDex/ApDexSettings.tsx:85]: same constant

File: frontend/src/pages/MetricsApplication/ApDex/types.ts [hunk 30]
```diff
@@ -0,0 +1,10 @@
+import { AxiosResponse } from 'axios';
+import { ApDexPayloadProps } from 'types/api/metrics/getApDex';
+
+export interface ApDexSettingsProps {
+	servicename: string;
+	handlePopOverClose: () => void;
+	isLoading?: boolean;
+	data?: AxiosResponse<ApDexPayloadProps[]> | undefined;
```
Comment (palashgdev) [inline frontend/src/pages/MetricsApplication/ApDex/types.ts:8]:
```suggestion
	data?: AxiosResponse<ApDexPayloadProps[]>;
```

File: frontend/src/pages/MetricsApplication/ApDex/ApDexSettings.tsx [hunk 31]
```diff
@@ -0,0 +1,115 @@
...
+}: ApDexSettingsProps): JSX.Element {
+	const [thresholdValue, setThresholdValue] = useState(() => {
+		if (data) {
+			return data.data[0].threshold;
+		}
+		return 0;
+	});
+	const { notifications } = useNotifications();
+
+	const { isLoading: setApDexIsLoading, mutateAsync } = useSetApDexSettings({
```
Comment (palashgdev) [inline frontend/src/pages/MetricsApplication/ApDex/ApDexSettings.tsx:37]:
```suggestion
	const { isLoading: isAppDexLoading, mutateAsync } = useSetApDexSettings({
```

File: frontend/src/container/GridPanelSwitch/utils.ts [hunk 32]
```diff
@@ -0,0 +1,22 @@
+import React, { ReactNode } from 'react';
+
+export const generateGridTitle = (title: ReactNode): string => {
```
Comment (palashgdev) [inline frontend/src/container/GridPanelSwitch/utils.ts:3]: export const generateGridTitle = (title: ReactNode): string => {
  if (React.isValidElement(title)) {
    return Array.isArray(title.props.children)
      ? title.props.children.map(child => typeof child === 'string' ? child : '').join(' ')
      : title.props.children;
  }
  
  return title?.toString() || '';
};

File: frontend/src/container/GridPanelSwitch/types.ts [hunk 33]
```diff
@@ -14,7 +15,7 @@ import { PANEL_TYPES } from '../../constants/queryBuilder';
 export type GridPanelSwitchProps = {
 	panelType: PANEL_TYPES;
 	data: ChartData;
-	title?: string;
+	title?: ReactNode;
```
Comment (palashgdev) [inline frontend/src/container/GridPanelSwitch/types.ts:18]: grab from widgets

File: frontend/src/pages/MetricsApplication/utils.ts [hunk 34]
```diff
@@ -15,3 +18,37 @@ export const getMetricsApplicationKey = (
 
 	return MetricsApplicationTab.OVER_METRICS;
 };
+
+export const onSaveApDexSettings = ({
```
Comment (palashgdev) [inline frontend/src/pages/MetricsApplication/utils.ts:22]:
```tsx
export const onSaveApDexSettings = ({
	thresholdValue,
	refetchGetApDexSetting,
	mutateAsync,
	notifications,
	handlePopOverClose,
	servicename,
}: OnSaveApDexSettingsProps) => async (): Promise<void> => {
	if (!refetchGetApDexSetting) return;

	try {
		await mutateAsync({
			servicename,
			threshold: thresholdValue,
			excludeStatusCode: '',
		});
		await refetchGetApDexSetting();
	} catch (err) {
		notifications.error({
			message: axios.isAxiosError(err) ? err.message : SOMETHING_WENT_WRONG,
		});
	} finally {
		handlePopOverClose();
	}
};
```

File: frontend/src/types/api/metrics/getApDex.ts [hunk 35]
```diff
@@ -0,0 +1,20 @@
+export type ApDexPayloadProps = {
+	serviceName: string;
+	threshold: number;
+	excludeStatusCodes: string;
+};
+
+export interface SetApDexSettingsProps {
```
Comment (palashgdev) [inline frontend/src/types/api/metrics/getApDex.ts:7]: duplicate props

File: frontend/src/pages/MetricsApplication/types.ts [hunk 36]
```diff
@@ -9,3 +16,16 @@ export const TAB_KEY_VS_LABEL = {
 	[MetricsApplicationTab.DB_CALL_METRICS]: 'DB Call Metrics',
 	[MetricsApplicationTab.EXTERNAL_METRICS]: 'External Metrics',
 };
+
+export interface OnSaveApDexSettingsProps {
+	thresholdValue: number;
+	servicename: string;
+	notifications: NotificationInstance;
+	refetchGetApDexSetting: (() => void) | undefined;
```
Comment (palashgdev) [inline frontend/src/pages/MetricsApplication/types.ts:24]:
```suggestion
	refetchGetApDexSetting?: VoidFunction;
```

File: frontend/src/pages/MetricsApplication/types.ts [hunk 37]
```diff
@@ -9,3 +16,16 @@ export const TAB_KEY_VS_LABEL = {
...
+	mutateAsync: UseMutateAsyncFunction<
+		SetApDexPayloadProps,
+		Error,
+		SetApDexSettingsProps
+	>;
+	handlePopOverClose: () => void;
```
Comment (palashgdev) [inline frontend/src/pages/MetricsApplication/types.ts:30]:
```suggestion
	handlePopOverClose: VoidFunction;
```

**Comment by palashgdev** [comment]
Screenshot 2023-08-12 at 11 06 39 AM
use Space from antd

**Comment by palashgdev** [comment]
use InputNumber from antd and handle wheelUp event

Screenshot 2023-08-12 at 11 09 01 AM

File: frontend/src/api/metrics/ApDex/getApDexSettings.ts (hunk 22 above)
Comment (Rajat-Dabade) [reply frontend/src/api/metrics/ApDex/getApDexSettings.ts:8]: Done.

File: frontend/src/container/MetricsApplication/MetricsPageQueries/OverviewQueries.ts (hunk 24 above)
Comment (Rajat-Dabade) [reply frontend/src/container/MetricsApplication/MetricsPageQueries/OverviewQueries.ts:189]: Okay.

File: frontend/src/container/MetricsApplication/MetricsPageQueries/OverviewQueries.ts (hunk 25 above)
Comment (Rajat-Dabade) [reply frontend/src/container/MetricsApplication/MetricsPageQueries/OverviewQueries.ts:236]: Removed it.

File: frontend/src/container/MetricsApplication/Tabs/Overview/ApDex/ApDexMetrics.tsx (hunk 26 above)
Comment (Rajat-Dabade) [reply frontend/src/container/MetricsApplication/Tabs/Overview/ApDex/ApDexMetrics.tsx:49]: Okay

File: frontend/src/pages/MetricsApplication/ApDex/ApDexApplication.tsx (hunk 27 above)
Comment (Rajat-Dabade) [reply frontend/src/pages/MetricsApplication/ApDex/ApDexApplication.tsx:12]: Okay.

File: frontend/src/pages/MetricsApplication/ApDex/ApDexSettings.tsx (hunk 28 above)
Comment (Rajat-Dabade) [reply frontend/src/pages/MetricsApplication/ApDex/ApDexSettings.tsx:1]: Sure.

File: frontend/src/pages/MetricsApplication/ApDex/ApDexSettings.tsx (hunk 29 above)
Comment (Rajat-Dabade) [reply frontend/src/pages/MetricsApplication/ApDex/ApDexSettings.tsx:85]: Okay.

File: frontend/src/pages/MetricsApplication/ApDex/types.ts (hunk 30 above)
Comment (Rajat-Dabade) [reply frontend/src/pages/MetricsApplication/ApDex/types.ts:8]: Done.

File: frontend/src/pages/MetricsApplication/ApDex/ApDexSettings.tsx (hunk 31 above)
Comment (Rajat-Dabade) [reply frontend/src/pages/MetricsApplication/ApDex/ApDexSettings.tsx:37]: Okay.

File: frontend/src/container/GridPanelSwitch/utils.ts (hunk 32 above)
Comment (Rajat-Dabade) [reply frontend/src/container/GridPanelSwitch/utils.ts:3]: Okay.

File: frontend/src/container/GridPanelSwitch/types.ts (hunk 33 above)
Comment (Rajat-Dabade) [reply frontend/src/container/GridPanelSwitch/types.ts:18]: Okay.

File: frontend/src/pages/MetricsApplication/utils.ts (hunk 34 above)
Comment (Rajat-Dabade) [reply frontend/src/pages/MetricsApplication/utils.ts:22]: Okay.

File: frontend/src/types/api/metrics/getApDex.ts (hunk 35 above)
Comment (Rajat-Dabade) [reply frontend/src/types/api/metrics/getApDex.ts:7]: Removed it.

File: frontend/src/pages/MetricsApplication/types.ts (hunk 36 above)
Comment (Rajat-Dabade) [reply frontend/src/pages/MetricsApplication/types.ts:24]: Okay.

File: frontend/src/pages/MetricsApplication/types.ts (hunk 37 above)
Comment (Rajat-Dabade) [reply frontend/src/pages/MetricsApplication/types.ts:30]: Okay.

File: frontend/src/container/GridGraphLayout/Graph/WidgetGraphComponent.tsx (hunk 23 above)
Comment (Rajat-Dabade) [reply frontend/src/container/GridGraphLayout/Graph/WidgetGraphComponent.tsx]: Done.

**Comment by srikanthccv** [comment]
I hope the points shared in the https://github.com/SigNoz/signoz/pull/3283#issuecomment-1674724736. Please test it yourself and let me know when it is ready for testing from my side.

**Review by palashgdev — APPROVED** [review]
LGTM

**Review by srikanthccv — APPROVED** [review]
@palashgdev @Rajat-Dabade I am approving this as the issue from the first part of video is not specific to this feature but a general chart issue.

**Pull request merged**